### PR TITLE
[ENH] Implement ConvTran deep learning estimator for classification and regression in PyTorch

### LINKED ISSUE #9465: [ENH] Implement ConvTran deep learning estimator for classification and regression in PyTorch
**Is your feature request related to a problem? Please describe.**
This is a sub-issue of 
Implement `ConvTran` classifier and regressor in `PyTorch`.

**Describe the solution you'd like**
ConvTran Paper: https://link.springer.com/article/10.1007/s10618-023-00948-2#Sec17
ConvTran GitHub: https://github.com/Navidfoumani/ConvTran

For implementation reference, refer to other PyTorch estimators implemented in `sktime/networks/`.
Especially, refer to the RNN implementation of PyTorch and the PR discussions to understand the design choices and potential improvements that can be made.
[SimpleRNNClassifierTorch](https://github.com/sktime/sktime/blob/main/sktime/classification/deep_learning/rnn/_rnn_torch.py) implemented in https://github.com/sktime/sktime/pull/8842 can be used as a template for classifier migrations. Similarly, [SimpleRNNRegressorTorch](https://github.com/sktime/sktime/blob/main/sktime/regression/deep_learning/rnn/_rnn_torch.py) implemented in https://github.com/sktime/sktime/pull/9013 can be used as a template for regressor migrations.

Please include the following checklist in your PR.

- [ ] Implemented the network structure in sktime/sktime/networks/
- [ ] Implemented the classifier in sktime/sktime/classification/deep_learning/
- [ ] Implementded the regressor in sktime/sktime/regressor/deep_learning/
- [ ] Followed the naming convention and directory structure from RNN migration samples
- [ ] Followed the [sktime docstring convention](https://www.sktime.net/en/latest/developer_guide/documentation.html#docstring-conventions). Please refer the RNN migration samples to understand the usage.
- [ ] Added meaningful test parameters in the classmethod get_test_params for classifiers and regressors.
- [ ] Followed the [sktime guide on implementing estimators](https://www.sktime.net/en/latest/developer_guide/add_estimators.html)
- [ ] Used same defaults in torch implemention as used in the corresponding tensorflow implementation of same model.

**Additional context**
- Migration issue from Tensorflow to PyTorch: #8699 
- https://github.com/sktime/sktime/issues/8928 consolidates further improvements/enhancements that are desirable and will be appreciated if you choose to incorporate. Feel free to brainstorm or share your ideas on these.

> Also explore forecasting and anomaly detection tasks using ConvTran

### PR DESCRIPTION
**Pull request by srupat**
#### Reference Issues/PRs
Fixes #9465

#### What does this implement? Explain your changes.
Implements the ConvTran regressor and classifier in PyTorch. 
ConvTran paper: https://link.springer.com/article/10.1007/s10618-023-00948-2
ConvTran GitHub: https://github.com/Navidfoumani/ConvTran

#### Does your contribution introduce a new dependency? If yes, which one?
No.

#### What should a reviewer concentrate their feedback on?
- The ConvTran paper mainly focuses on two important things, implementing a new scheme for absolute position encoding and implementing a new scheme for relative position encoding. Refer to `_positional_encoding.py` and `_attention.py` respectively.
- Base ConvTran model in `networks/convtran/_convtran_torch.py`
- The ConvTran classifier and regressor implemented in PyTorch.

#### Did you add any tests for the change?
No.

#### Any other comments?
I have followed all the considerations and maintainer reviews while implementing this estimator. Refer to PR #9412 

##### For all contributions
- [x] Optionally, for added estimators: I've added myself and possibly to the `maintainers` tag - do this if you want to become the owner or maintainer of an estimator you added.
  See here for further details on the [algorithm maintainer role](https://www.sktime.net/en/latest/get_involved/governance.html#algorithm-maintainers).
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

##### For new estimators
- [x] I've added the estimator to the API reference - in `docs/source/api_reference/taskname.rst`, follow the pattern.
- [x] I've added one or more illustrative usage examples to the docstring, in a pydocstyle compliant `Examples` section.

##### Specific to DL-based estimators
- [x] Implemented the network structure in sktime/sktime/networks/
- [x] Implemented the classifier in sktime/sktime/classification/deep_learning/
- [x] Implementded the regressor in sktime/sktime/regressor/deep_learning/
- [x] Followed the naming convention and directory structure from RNN migration samples
- [x] Followed the [sktime docstring convention](https://www.sktime.net/en/latest/developer_guide/documentation.html#docstring-conventions). Please refer the RNN migration samples to understand the usage.
- [x] Added meaningful test parameters in the classmethod `get_test_params` for classifiers and regressors.
- [x] Followed the [sktime guide on implementing estimators](https://www.sktime.net/en/latest/developer_guide/add_estimators.html)

#### Additional Context
- TapNet PR on migration of TapNet estimator from tensorflow to torch - #9412

### CODE REVIEW

**Comment by srupat** [comment]
@RecreationalMath could you please review this PR? Thanks!

File: sktime/networks/convtran/_attention.py [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+            self.query(x)
+            .reshape(batch_size, seq_len, self.num_heads, -1)
+            .transpose(1, 2)
+        )
+
+        torch_matmul = self._torch_op("torch.matmul")
+        attn = torch_matmul(q, k) * self.scale
+        softmax = self._torch_op("torch.nn.functional.softmax")
+        attn = softmax(attn, dim=-1)
+
```
[the remark this answers is not in the document]
Comment (srupat) [reply sktime/networks/convtran/_attention.py:209]: This is intentional. The [paper](https://link.springer.com/article/10.1007/s10618-023-00948-2) explicitly mentions adding the relative bias _after_ the softmax function, as they empirically found out that attention models perform better with this approach.
**Refer to section 4.2 Efficient relative position encoding (eRPE) of the paper.**

File: sktime/networks/convtran/_positional_encoding.py [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+        x : torch.Tensor of shape (batch_size, seq_len, d_model)
+            Input tensor containing embedded time series.
+
+        Returns
+        -------
+        torch.Tensor of shape (batch_size, seq_len, d_model)
+            Tensor with positional encodings added.
+        """
+        x = x + self.pe
+        return self.dropout(x)
```
[the remark this answers is not in the document]
Comment (srupat) [reply sktime/networks/convtran/_positional_encoding.py]: Fixed in 55daafd

File: sktime/networks/convtran/_positional_encoding.py [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+        ----------
+        x : torch.Tensor of shape (batch_size, seq_len, d_model)
+            Input tensor containing embedded time series.
+
+        Returns
+        -------
+        torch.Tensor of shape (batch_size, seq_len, d_model)
+            Tensor with positional encodings added.
+        """
+        x = x + self.pe
```
[the remark this answers is not in the document]
Comment (srupat) [reply sktime/networks/convtran/_positional_encoding.py]: Fixed in 55daafd

File: sktime/networks/convtran/_convtran_torch.py [hunk 4]
```diff
@@ -0,0 +1,563 @@
...
+        out = att + self.feed_forward(att)
+        out = self.layer_norm_2(out)
+
+        out = out.permute(0, 2, 1)
+        out = self.gap(out)
+        out = self.flatten(out)
+        out = self.out(out)
+        if self._activation_out is not None:
+            out = self._activation_out(out)
+        return out
```
[the remark this answers is not in the document]
Comment (srupat) [reply sktime/networks/convtran/_convtran_torch.py]: Fixed in 55daafd

File: sktime/networks/convtran/_positional_encoding.py [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+        return self.dropout(x)
```
[the remark this answers is not in the document]
Comment (srupat) [reply sktime/networks/convtran/_positional_encoding.py]: Fixed in 55daafd

**Comment by srupat** [comment]
@fkiraly Hi, just bumping this PR as it is ready for review and passes all CI checks. Thanks!
